Use REQUIRES_FULL_LIBC in tests and samples instead of NEWLIB_LIBC

keith-packard opened the pull request:
To gain broader test coverage for the default C library, convert all samples and tests from specifying newlib to specifying REQUIRES_FULL_LIBC. Environments that require newlib will still use it, but otherwise these tests will build with picolibc.

keith-packard committed:
libc: Add REQUIRES_FLOAT_PRINTF to indirectly set printf support

Instead of making applications use C library specific settings to enable
floating point support in printf, provide this indirect symbol which then
detects which C library is in use and selects the correct configuration for
each.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
samples: Switch from NEWLIB_LIBC to REQUIRES_FULL_LIBC

Instead of forcing use of NEWLIB_LIBC, select any available complete C
library implementation. Add CONFIG_REQUIRES_FLOAT_PRINTF and adjust
CONFIG_LIBC_MALLOC_ARENA_SIZE as needed.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
tests: Switch from NEWLIB_LIBC to REQUIRES_FULL_LIBC

Instead of forcing use of NEWLIB_LIBC, select any available complete C
library implementation. Add CONFIG_REQUIRES_FLOAT_PRINTF where needed.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
This is stack on top of the series that disables malloc-by-default in picolibc #63259

jukkar commented on subsys/net/lib/sockets/Kconfig:
```diff
@@ -6,6 +6,7 @@
 menuconfig NET_SOCKETS
```
At least for networking commits, there should be no need to mention subsys/ in commit subject. Also there is no sockets subsystem, just networking one.
I suggest just say `net/sockets: Add REQUIRES_MALLOC when enabling sockets`

keith-packard replied:
Thanks. Fixed.

jukkar commented on samples/net/sockets/big_http_download/src/big_http_download.c:
```diff
@@ -13,13 +13,14 @@
 
```
For the commit `samples/sockets: Fix native app detection when selecting headers`, I suggest to change the subject to
`samples/net/sockets: ....`
so that the searching networking commits gets slightly easier.

keith-packard replied:
Sounds good. Fixed.

jukkar reviewed (commented):
Networking related changes look ok, couple of minor nits about the commit messages.

jhedberg commented:
@keith-packard do you think this should go into the 3.5 release?

fabiobaltieri commented:
Hey @keith-packard I've pushed this up to https://github.com/zephyrproject-rtos/zephyr-testing/tree/vfabio-testing-branch to get a push run and got few failures: https://github.com/zephyrproject-rtos/zephyr-testing/runs/17600258307 could you have a look into those? It should be equivalent of the main run we get after merge and I don't think those were failing before, but it may just be new combinations that were not previously run.

keith-packard commented:
> update: `west build -p -b native_posix -T samples/posix/gettimeofday/sample.posix.gettimeofday` fails on main too so it's unrelated, not sure about lz4 ones

The gettimeofday failure is fixed by #63331 

The lz4 failures look like they're caused by having too small a default malloc heap. I've adjust the size of that in the prj.conf file for that test. We'll see if that makes it happy.

fabiobaltieri commented:
> The gettimeofday failure is fixed by #63331

Thanks!

> The lz4 failures look like they're caused by having too small a default malloc heap. I've adjust the size of that in the prj.conf file for that test. We'll see if that makes it happy.

Yeah good call, https://github.com/zephyrproject-rtos/zephyr-testing/runs/17615813727 they are gone.

Not sure how I feel about merging this now, we have one more full run before the release, it'd be nice to get the extra coverage, I fear there may be more odd stuff broken but then if it's actual problems it may not be a bad thing. @jhedberg what do you think?

fabiobaltieri commented:
> > The gettimeofday failure is fixed by #63331

Actually one more about this one, is this error exposed by the change in that PR? Unclear to me why the current push run does not fail.

keith-packard commented:
> > > The gettimeofday failure is fixed by #63331
> 
> Actually one more about this one, is this error exposed by the change in that PR? Unclear to me why the current push run does not fail.

no, that PR fixes the bug -- I haven't explored why things weren't broken before, but with gcc -m32 on x86 systems, the compiler complains when you shift a time_t value by 32 bits -- time_t is only 32 bits in that configuration unless you tell the compiler to use 64 bits (which is what #63331 does).

keith-packard commented:
> Not sure how I feel about merging this now, we have one more full run before the release, it'd be nice to get the extra coverage, I fear there may be more odd stuff broken but then if it's actual problems it may not be a bad thing. @jhedberg what do you think?

Yup, it might be too close to the release for this series.

Hrm. Maybe what we should be doing is pulling the one actual fix out of this series and merging it separately? https://github.com/zephyrproject-rtos/zephyr/pull/63332/commits/b2ff1607be9edbac4665e8dc4d3d3a9ba1f0428a is a small configuration fix to a driver that allows it to be used with either newlib or picolibc. I'd left it in this series because that way it gets tested, but it could be applied separately.

fabiobaltieri commented:
> Hrm. Maybe what we should be doing is pulling the one actual fix out of this series and merging it separately? [b2ff160](https://github.com/zephyrproject-rtos/zephyr/commit/b2ff1607be9edbac4665e8dc4d3d3a9ba1f0428a) is a small configuration fix to a driver that allows it to be used with either newlib or picolibc. I'd left it in this series because that way it gets tested, but it could be applied separately.

Yeah that should certainly go in if you are ok splitting it out and keeping the rest for 3.6.

jhedberg commented:
> > Hrm. Maybe what we should be doing is pulling the one actual fix out of this series and merging it separately? [b2ff160](https://github.com/zephyrproject-rtos/zephyr/commit/b2ff1607be9edbac4665e8dc4d3d3a9ba1f0428a) is a small configuration fix to a driver that allows it to be used with either newlib or picolibc. I'd left it in this series because that way it gets tested, but it could be applied separately.
> 
> Yeah that should certainly go in if you are ok splitting it out and keeping the rest for 3.6.

Agreed

keith-packard commented:
> Agreed

#63827

aescolar commented:
> update: `west build -p -b native_posix -T samples/posix/gettimeofday/sample.posix.gettimeofday` fails on main too so it's unrelated

`samples/posix/gettimeofday/sample.posix.gettimeofday` is not compatible with native_posix

aescolar reviewed (changes requested):
The overall direction of the PR wrt. NEWLIBC->REQUIRES_FULL_LIBC is good , but there is a few things that need fixing.

A sidenote: note that for native_sim the default library was never changed to picolibc and remains the host library

aescolar commented on samples/net/sockets/echo_async_select/sample.yaml:
```diff
@@ -7,7 +7,7 @@ common:
     - net
     - socket
-  filter: TOOLCHAIN_HAS_NEWLIB == 1
+  filter: CONFIG_FULL_LIBC_SUPPORTED
```
Side-tangent: the extra_configs in sample.net.sockets.echo_async_select.posix is a bit bogus as CONFIG_POSIX_API=y is set in the prj.conf. The whole thing needs a filter on ` !CONFIG_NATIVE_APPLICATION`

aescolar commented on samples/posix/eventfd/sample.yaml:
```diff
@@ -2,7 +2,7 @@ sample:
   name: eventfd
 common:
-  filter: TOOLCHAIN_HAS_NEWLIB == 1
+  filter: CONFIG_FULL_LIBC_SUPPORTED
```
we need a  `and !CONFIG_NATIVE_APPLICATION` filter

keith-packard commented on samples/net/sockets/big_http_download/src/big_http_download.c:
```diff
@@ -13,13 +13,14 @@
 #include "mbedtls/md.h"
 
-#if !defined(__ZEPHYR__) || defined(CONFIG_POSIX_API)
+#if defined(CONFIG_NATIVE_APPLICATION) || defined(CONFIG_POSIX_API)
```
I'll remove this patch from the series and see what things look like, and if things are still broken, figure out what changes are necessary to make things at least build.

keith-packard commented on tests/lib/cmsis_dsp/statistics/testcase.yaml:
```diff
@@ -12,7 +12,7 @@ tests:
   libraries.cmsis_dsp.statistics.fpu:
     filter: ((CONFIG_CPU_AARCH32_CORTEX_R or CONFIG_CPU_CORTEX_M) and CONFIG_CPU_HAS_FPU
-      and TOOLCHAIN_HAS_NEWLIB == 1) or CONFIG_ARCH_POSIX
+      and (CONFIG_FULL_LIBC_SUPPORTED or TOOLCHAIN_HAS_PICOLIBC )) or CONFIG_ARCH_POSIX
```
Hrm. Not sure how I messed this up, sorry for the noise.

keith-packard commented on tests/lib/hash_map/testcase.yaml:
```diff
@@ -19,8 +19,7 @@ tests:
-    filter: TOOLCHAIN_HAS_NEWLIB == 1
+    filter: CONFIG_FULL_LIBC_SUPPORTED
     extra_configs:
       - CONFIG_NEWLIB_LIBC=y
```
Yup. I've done another scan to find all users of CONFIG_NEWLIB_LIBC=y or TOOLCHAIN_HAS_NEWLIB == 1 and I think I've got them all fixed now. There are still a handful of newlib-specific tests which use these, of course.

keith-packard commented on tests/lib/thrift/ThriftTest/testcase.yaml:
```diff
@@ -5,7 +5,7 @@ common:
     - newlib
```
It should be. I've checked all uses of the newlib tag in tests (there aren't any in samples).

keith-packard commented on samples/posix/gettimeofday/sample.yaml:
```diff
@@ -2,7 +2,7 @@ sample:
   name: gettimeofday
 common:
-  filter: ( TOOLCHAIN_HAS_NEWLIB == 1 and not CONFIG_SOC_FAMILY_INTEL_ADSP )
+  filter: ( CONFIG_FULL_LIBC_SUPPORTED and not CONFIG_SOC_FAMILY_INTEL_ADSP )
```
ah, that's how we keep this test from trying to build with the host C library. Awesome.

keith-packard commented on samples/net/sockets/dumb_http_server/sample.yaml:
```diff
@@ -2,7 +2,7 @@ sample:
   name: socket_dumb_http_server
 common:
-  filter: TOOLCHAIN_HAS_NEWLIB == 1
+  filter: CONFIG_FULL_LIBC_SUPPORTED
```
Sounds like this would get handled by #63855 ? So using !CONFIG_EXTERNAL_LIBC should be sufficient to avoid building this for native_posix?

keith-packard commented on samples/net/sockets/http_get/sample.yaml:
```diff
@@ -2,7 +2,7 @@ sample:
   name: socket_http_get
 common:
-  filter: TOOLCHAIN_HAS_NEWLIB == 1
+  filter: CONFIG_FULL_LIBC_SUPPORTED
```
should this be `and not CONFIG_EXTERNAL_LIBC` instead?

aescolar replied:
`not NATIVE_LIBC` would be good yes.

keith-packard commented:
ok, I'm seeing a fairly persistent pattern here -- a bunch of these samples and tests cannot be used with the host C library, even though that library provides a full C library implementation. So the filters on `TOOLCHAIN_HAS_NEWLIB == 1` and the setting of `CONFIG_NEWLIB_LIBC=y` were performing two different purposes:

 1. Building with newlib instead of the minimal C library on embedded targets
 2. Skipping tests/samples when using the host C library.

I think we need a new config flag for the second test here -- we need to know when the host C library is in use and filter tests in that case. I think that's equivalent to `NATIVE_APPLICATION or (NATIVE_LIBRARY and not EXTERNAL_LIBC)` -- arch_posix uses `EXTERNAL_LIBC`  internally to drive selection of the host C library.

keith-packard committed:
Kconfig.zephyr: Add NATIVE_LIBC and NATIVE_LIBCPP

These are set when building against the host C or C++ libraries. This
allows filtering tests that require use of a C library that sits atop
Zephyr OS APIs.

Signed-off-by: Keith Packard <keithp@keithp.com>

aescolar commented:
> I think we need a new config flag for the second test here -- we need to know when the host C library is in use and filter tests in that case. I think that's equivalent to `NATIVE_APPLICATION or (NATIVE_LIBRARY and not EXTERNAL_LIBC)` -- arch_posix uses `EXTERNAL_LIBC` internally to drive selection of the host C library.

I'm in favor of the NATIVE_LIBC & NATIVE_LIBCPP you are introducing in this PR.
EXTERNAL_LIBC was introduced as a synonym of the host libC for the native target, and in vanilla Zephyr that's all it was, but it has always been a misnomer, and somebody may have used it downstream to select an actual external C library for embedded targets.

aescolar commented on tests/lib/thrift/ThriftTest/testcase.yaml:
```diff
@@ -5,7 +5,7 @@ common:
   modules:
     - thrift
-  filter: TOOLCHAIN_HAS_NEWLIB == 1
+  filter: CONFIG_FULL_LIBC_SUPPORTED
```
We can also just let it be
The platform_allow filters it. Though platform_allow is unfortunately used as a "prefered platforms" filter in many cases.

aescolar commented on samples/basic/hash_map/sample.yaml:
```diff
@@ -36,21 +36,21 @@ tests:
-    filter: TOOLCHAIN_HAS_NEWLIB == 1
+    filter: CONFIG_FULL_LIBC_SUPPORTED
     extra_configs:
       - CONFIG_NEWLIB_LIBC=y
```
These I really think they should be left 3, TOOLCHAIN_HAS_NEWLIB == 1
Otherwise twister will build them and if newlibc is not supported kconfig will warn and they will be built with whatever they support. That may make the test pass, but it is meaningless, as these tests seem specific for newlib.

keith-packard replied:
Thanks. Fixed.

aescolar commented on samples/net/cloud/aws_iot_mqtt/sample.yaml:
```diff
@@ -4,7 +4,7 @@ sample:
   tags: net mqtt cloud
   harness: net
-  filter: TOOLCHAIN_HAS_NEWLIB == 1
+  filter: CONFIG_FULL_LIBC_SUPPORTED
```
note in this case, platform allow is abused as a "platform preferred". This sample builds just fine for lots of other platforms, so I would add the extra `and !CONFIG_NATIVE_LIBC`

aescolar commented:
@keith-packard thanks for the fixes. Note the 2 comments I left unresolved, I think those should still be fixed.

aescolar commented:
@keith-packard how is this PR looking?
It would be nice to get this fixes in.
Specially https://github.com/zephyrproject-rtos/zephyr/pull/63332/commits/58f4b31f0a8e4096b79e2b6edf39f5fdb13ab0df would help on other PRs.

keith-packard commented:
> @keith-packard how is this PR looking? It would be nice to get this changes fixes in. Specially [58f4b31](https://github.com/zephyrproject-rtos/zephyr/commit/58f4b31f0a8e4096b79e2b6edf39f5fdb13ab0df) would help on other PRs.

I've rebased it this morning; there were conflicts with the removal of `CONFIG_ZTEST_NEW_API` which I needed to fix up. Other than that, I think it's ready to go. Let's see if CI is happy after the rebase.

aescolar commented:
@keith-packard 
Thanks, but please don't forget https://github.com/zephyrproject-rtos/zephyr/pull/63332/files#diff-d9b05ecb84aaf9fb90748a2acbe37e8bf483c7a3c1370044dde783af45a1ab37

fabiobaltieri commented:
@keith-packard could you rebase please?
